viewer/src/main/webapp/viewer-html/components/OpenLayersSnappingController.js

```diff
@@ -0,0 +1,259 @@
+            // assume we now have a drawing or editing layer
+            this.frameworkLayer = options.layer.getFrameworkLayer()
+            this.frameworkControl.setLayer(this.frameworkLayer);
+            this.activate();
+        }
+    },
+    /**
+     * add the snapping target.
+     * @param {type} snappingLayer
+     */
+    addLayerDataFor: function (appLayer) {
+        var me = this;
+        // lookup feature source
+        var featureService = this.config.viewerController.getAppLayerFeatureService(appLayer);
+        // find geom attribute
+        featureService.loadAttributes(appLayer, function (result) {
+            var geomAttribute = appLayer.attributes[appLayer.geometryAttributeIndex].name;
+            var extent = me.viewerController.mapComponent.getMap().getExtent();
+            // fetch/load geometries
+            featureService.loadFeatures(
+                    appLayer,
+                    me.parseFeatures,
+                    undefined, {
+                        store: 1,
+                        limit: 1000,
+                        arrays: 1,
+                        // just get geometry
+                        attributesToInclude: [geomAttribute],
+                        edit: false,
+                        graph: true,
+                        // only for map extent
+                        filter: "INTERSECTS(" + geomAttribute + ", " + extent.toWKT() + ")"
+                    }, {
+                /* we need access to appLayer.id and this in processing the response */
+                me: me,
+                appLayer: appLayer
+            });
+        });
+    },
+    /**
+     * remove the snapping target.
+     * @param {type} snappingLayer
+     */
+    removeLayer: function (appLayer) {
+        this.deactivate();
+        //look up snappingLayer primitive by name/id...
+        var rLyr = this.frameworkMap.getLayersByName(this.getlayerName(appLayer));
+        // there should only be one layer in the rLyr
+        rLyr = rLyr[0];
+        Ext.Array.remove(this.snapLayers, rLyr);
+        this.frameworkMap.removeLayer(rLyr);
+        this.frameworkControl.removeTargetLayer(rLyr);
+        this.activate();
+    },
+    /**
+     * remove all snapping targets.
+     */
+    removeAll: function () {
+        for (var i = 0; i < this.snapLayers.length; i++) {
+            this.frameworkControl.removeTargetLayer(this.snapLayers[i]);
+            this.frameworkMap.removeLayer(this.snapLayers[i]);
+        }
+        this.snapLayers = [];
+        this.deactivate();
+    },
+    /**
+     * activate snapping, if there are snapping targets
+     */
+    activate: function () {
+        if (this.snapLayers.length > 0) {
+            this.frameworkControl.activate();
+        }
+    },
+    /**
+     * deactivate snapping.
+     */
+    deactivate: function () {
+        this.frameworkControl.deactivate();
+    },
+    parseFeatures: function (data) {
```

Comment:
Is it fast enough? Can imagine when having large datasets (or a zoomed out extent), a lot of features are being retrieved and shown.
Client side caching?